okhttp/src/jvmMain/kotlin/okhttp3/Dispatcher.kt

```diff
@@ -164,12 +196,27 @@ class Dispatcher() {
     val executableCalls = mutableListOf<AsyncCall>()
     val isRunning: Boolean
     synchronized(this) {
-      val i = readyAsyncCalls.iterator()
-      while (i.hasNext()) {
-        val asyncCall = i.next()
+      while (readyAsyncCalls.isNotEmpty()) {
+        val filteredCalls = readyAsyncCalls.mapNotNull {
+          if (it.callsPerHost.get() < maxRequestsPerHost) it.call else null
+        }
+
+        if (filteredCalls.isEmpty()) break // Max host capacity for all requests.
+
+        val nextCall = asyncCallDispatcher.nextCall(filteredCalls)
```

Comment:
I worry about putting user logic in this hot path.  Potentially that needs to process through large list of items, on every pass through promoteAndExecute.  I wonder whether a more constrained version without the same risk, is to return a priority for each item, with a default returning 0 for all.  These have a priority queue for readyAsyncCalls.

Reply:
I agree with you, but isn't returning priority for each item essentially the same as just returning next call to execute? Determining the priority takes just as long as just returning the call.

Reply:
It qould mean you could do it once on insertion. Or is dynamic priority important here?

Reply:
Yes, at least for my use case. I am trying to load a bunch of images and prioritize images that are visible to the user and that requires dynamic priority. Maybe we can combine both methods to get an optimal result for everyone?

Reply:
Hmm, how about running `promoteAndExecute` on another thread? (possibly using coroutines)  
running this every time user calls `enqueue` sounds quite problematic.

Reply:
I'd suggest waiting for other input, particularly @swankjesse. The more we tweak this the more frustrating it will be if it requires a different approach.